src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -369,6 +469,30 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
                 throw Error.ArgumentNull("resourceWrapper");
             }
 
+            if (resourceWrapper.Resource.Id != null)
+            {
+                IWebApiRequestMessage request = readContext.InternalRequest;
+                IWebApiUrlHelper urlHelper = readContext.InternalUrlHelper;
+
+                DefaultODataPathHandler pathHandler = new DefaultODataPathHandler();
+                string serviceRoot = urlHelper.CreateODataLink(
+                    request.Context.RouteName,
+                    request.PathHandler,
+                    new List<ODataPathSegment>());
+
+                ODataPath odataPath = pathHandler.Parse(serviceRoot, resourceWrapper.Resource.Id.ToString(), request.RequestContainer);
+                KeySegment keys = odataPath.Segments.OfType<KeySegment>().FirstOrDefault();
+
+                if (keys == null)
+                {
+                    throw new InvalidOperationException("The link does not contain a key.");
+                }
+
+                var pk = keys.Keys.First().Key;
+                var pkv = keys.Keys.First().Value;
```

Comment:
I think calling `keys.Keys.First()` twice is wasteful. May be change to:
```
var keyValPair = keys.Keys.First();
keyValPair.Key;
keyValPair.Value;
```

Reply:
Also, why only use the first key? We don't support composite keys for this feature?